### PR TITLE
[feat, UX] Add external link to Wallabag

### PR DESCRIPTION
**Pull request by Frenzie**
Offers a choice when Wallabag is available.

## Android/desktop (with Wallabag enabled)
Screenshot_2019-03-31_15-39-29

## Other platforms (with Wallabag enabled)
Screenshot_2019-03-31_15-56-03

### CODE REVIEW

File: frontend/apps/reader/modules/readerlink.lua [hunk 1]
```diff
@@ -576,56 +628,60 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
             })
         end
         return true
-    end
+    elseif self.ui.wallabag then
+        local settings = G_reader_settings:readSetting("external_link_action")
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerlink.lua]: Forgot to add the menu item to set this setting? Or just undocumented for now?
(`local settings =` plural seems like a bad name for that variable :)

File: frontend/apps/reader/modules/readerlink.lua [hunk 2]
```diff
@@ -576,56 +628,60 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
...
-        local provider = DocumentRegistry:getProvider(linked_filename)
-        if provider then
-            -- Display filename with anchor or query string, so the user gets
-            -- this information and can manually go to the appropriate place
-            local display_filename = linked_filename
-            if anchor then
-                display_filename = display_filename .. anchor
+                },
+            }
+            if Device:openLink() == nil then
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerlink.lua]: Calling `Device:openLink()` to know if it's supported seems a bit twisted. It returns `false `when set to `no`, and returns `nil` when it is supported, but provided with some invalid input...
If there's no cleaner way, at least `if Device:openLink() ~= false then` ?

File: frontend/apps/reader/modules/readerlink.lua [hunk 3]
```diff
@@ -576,56 +628,60 @@ function ReaderLink:onGotoLink(link, neglect_current_location, allow_footnote_po
...
-                    end)
-                end
-            })
-        else
-            UIManager:show(InfoMessage:new{
-                text = T(_("Link to unsupported local file:\n%1"), link_url),
-            })
+
+            choose_action = ButtonDialogTitle:new{
+                title = T(_("External link:\n\n'%1'"), link_url),
```
Comment (poire-z) [inline frontend/apps/reader/modules/readerlink.lua]: Not fond of the url inside `' '` on the screenshots. Can't you just let it plain?

File: frontend/apps/reader/modules/readerlink.lua (hunk 1 above)
Comment (Frenzie) [reply frontend/apps/reader/modules/readerlink.lua]: I copied this from ReaderStatus. I wouldn't have put the settings in yet if I'd written it from scratch.

File: frontend/apps/reader/modules/readerlink.lua (hunk 2 above)
Comment (Frenzie) [reply frontend/apps/reader/modules/readerlink.lua]: Not defining it as `no` and testing `Device.openLink`?

Comment (poire-z) [reply frontend/apps/reader/modules/readerlink.lua]: Dunno. Is there already functions not defined in generic/device.lua that are defined in others?
I thought the idea of them being `no` was some kind of API listing everything that can be available.
Or add Device:canOpenLink = yes|no ?

Comment (Frenzie) [reply frontend/apps/reader/modules/readerlink.lua]: I think it's mainly that you can safely call it and be sure it won't crash. API listing is more like a happy side effect.

Having a separate canOpenLink seems a touch inelegant but I suppose that'll have to be the way forward. :-)

**Comment by Frenzie** [comment]
I added a new menu item to the Links submenu. Looking at the screenshots below I realized it might be redundant with that ignore external links thing.

Screenshot_2019-03-31_17-23-18
Screenshot_2019-03-31_17-23-27

**Comment by poire-z** [comment]
> I added a new menu item to the Links submenu. Looking at the screenshots below I realized it might be redundant with that ignore external links thing.

I think it should be just below `Ignore external links` (as the swipe gestures currently ignore external links).
Or merged with it (but then "Ignore" instead of "Do nothing").
But on my Kobo with wallabag not enabled, it just serves as ignoring Wikipedia article links in Wikipedia EPUB for easier page change with Tap.
And ideally, the menu item should show the action selected, ie: `External links: ignore >` :)

**Comment by Frenzie** [comment]
> I think it should be just below Ignore external links (as the swipe gestures currently ignore external links).

Speaking of which, why is that so immensely complicated?

**Comment by Frenzie** [comment]
Screenshot_2019-03-31_17-51-39
Screenshot_2019-03-31_17-53-11

**Comment by poire-z** [comment]
> why is that so immensely complicated

It's not: you just haven't done it yet, and I just haven't requested yet it to be a toggable :) and probably another one from the Tap one - as it involves a bit of expensive work in crengine, but all the work in crengine/cre.cpp is already done, you just have to make that hardcoded `true` a variable:
https://github.com/koreader/koreader/blob/7396abebf26ecbb3a8cf45f4ee4dcb61cb788287/frontend/apps/reader/modules/readerlink.lua#L740-L747
and a bit of tweak at the end of `ReaderLink:onGoToPageLink()` to not do the xpointer checking when there is none but a `uri`.

**Comment by Frenzie** [comment]
I'm not really sure what you're asking tbh. I don't really understand the rationale behind what that setting does on closer examination. This probably comes back to my lack of understanding for why swipe to follow links is ignoring links.

**Comment by poire-z** [comment]
In Wikipedia EPUB, small footnotes are hard to tap on, and may be sitting just between 2 wikipedia URL links. Chances of taping ot swiping on it would be very low. Having avoid external links available makes that a non-issue.

**Comment by Frenzie** [comment]
I see. :-)

Anyhow, I think this PR now includes a response to all comments? (Other than dynamic menu text.)

**Comment by poire-z** [comment]
May be give that `External link action` some enabled_func like `if self.ui.document.info.has_pages or not isTapIgnoreExternalLinksEnabled`?

Otherwise lgtm.
(Except that now, that menu has 11 items and spans 2 pages :) I'll probably kill "Swipe to follow first link on page" as I don't see why one would use it since we have "Swipe to follow nearest link").

**Comment by Frenzie** [comment]
vcfrtdggggggggg ← that's my cat's response

**Comment by poire-z** [comment]
I've been looking at this to see how we could merge `Ignore external links` into the added `External link action>` submenu (and adding the selected action text into the main menu text).
I wondered if we should keep both "ignore" (let the tap propagate) and "do nothing" (eat the tap, don't propagate).

But after some time, I just wonder if this menu is really useful with all the actions.
I guess most will choose only "ignore" or "ask with popup" - just to see the url before getting into the wallabag wifi popup and download, or launching an external browser. These are just one tap away anyway.

We could be just fine with the existing single checkbox `Ignore external links`: if checked, let the tap propagate to change page or find an internal link near - if unchecked, just show the popup with the available actions?

**Comment by Frenzie** [comment]
I'd never choose "nothing" and I don't think add to Wallabag makes all that much sense as a default, but I can definitely see opting for "open in browser" without disabling the Wallabag plugin.

PS It's only got this configuration because you requested it! :-P

**Comment by poire-z** [comment]
>> Forgot to add the menu item to set this setting? Or just undocumented for now?

> I copied this from ReaderStatus

So, we have that menu because of that setting that just got there because it came along in the copying ? :)

**Comment by poire-z** [comment]
Or if we were to keep it, it may be better to have it as a real submenu with various combinable items, instead of an alternative single choice one.
External links>
- Ignore on tap
- Ignore on swipe
- Open in browser without asking
- Open Wikipedia links also in browser without asking (avoid our preview in textboxwidget if you prefer your browser)

**Comment by Frenzie** [comment]
> So, we have that menu because of that setting that just got there because it came along in the copying ? :)

Well no, but I wouldn't have added a menu for it this weekend because studying the menu and the code related to it is an entirely different task. :-P The open in browser setting is obvious but not terribly relevant on most platforms, the "nothing" setting is the kind of thing you like (and easy), and add to Wallabag is the reason I added this in the first place so it would be odd if it weren't a setting. It's the menu you get if you want a menu now rather than next month. ;-)

> it may be better to have it as a real submenu with various combinable items, instead of an alternative single choice one.

Naturally. ;-)

**Pull request merged**